File: C++/21.状态模式/StatePatternDemo.cpp
/**
 * @file StatePatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-22
 *
 * @copyright Copyright (c) 2019
 *
 */

#include <iostream>

#include "Context.h"
#include "StartState.h"
#include "StopState.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<Context> context = std::make_shared<Context>();

    std::shared_ptr<StartState> start_state = std::make_shared<StartState>();
    start_state->DoAction(context);

    std::cout << context->GetState()->ToString() << std::endl;

    std::shared_ptr<StopState> stop_state = std::make_shared<StopState>();
    stop_state->DoAction(context);

    std::cout << context->GetState()->ToString() << std::endl;
    return 0;
}

// g++ -std=c++11 -g StatePatternDemo.cpp StartState.cpp StopState.cpp Context.cpp -o StatePatternDemo

File: Objective-C/21.StatePattern/StatePatternDemo.m
//
//  StatePatternDemo.m
//  21.StatePattern
//
//  Created by Layne on 2020/11/19.
//

#import "Context.h"
#import "StartState.h"
#import "StopState.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        Context* context = [[Context alloc] init];

        StartState* startState = [[StartState alloc] init];
        [startState doAction:context];

        NSLog(@"%@", [context getState]);

        StopState* stopState = [[StopState alloc] init];
        [stopState doAction:context];

        NSLog(@"%@", [context getState]);
    }
    return 0;
}
